### ISSUE TITLE
Warning about primary MS run format

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#4332]
Since #4210 (@timosachsenberg), I get warnings like this when I run NucleicAcidSearchEngine:
```
To ensure tracability of results please prefer mzML files as primary MS run.
Filename: 'file:///X:\win\path/file.raw'
```
(Also note the mismatched path separators.)

The code for this is here: https://github.com/OpenMS/OpenMS/blob/5d961be216823625676053b6782fc385ec413c2c/src/openms/source/METADATA/ProteinIdentification.cpp#L376-L380

From our perspective, the .raw file *is* the primary MS file. We may delete converted mzML files, but will definitely keep the .raw files. I would expect that to be quite a common practice, too (among labs that use mzML files).

What is the point of this warning?
Shouldn't we be happy with .raw files as primary MS files?

**Comment by timosachsenberg** [comment]
There are a few reasons so I only list some of them here:
- some formats and downstream tools want to trace down a file that actually exists on the file system (e.g., for visualization) form the information contained in a e.g. mzTab file. This was the preferred way to carry readable spectra information through the other file formats.
- entry for the raw file is often malformed or not filled out - conversion from Thermo is kind of an exception
- folder often only exists on the instrument machine etc.

TL;DR to do meaningful visualize a result file (e.g., quant + id) at the end of a processing workflow we need the first "sensible" mzML files (e.g., after conversion or peak picking)

One could think about carrying the raw file name (if it exists - see comment above) in another meta data field.

**Comment by hendrikweisser** [comment]
Thanks for the quick reply, Timo, but I still find this problematic. Imagine an analysis pipeline that produces several intermediary files (e.g. filtering, signal processing):
.raw -> .mzML -> ... -> .mzML -> idXML/featureXML

Generally, I think it's most important to store the identity of the very first file (the .raw file), because all others are derived from that and it may provide the link to information about the experiment/sample.

For visualisation, you probably want the most recent .mzML file, not the first one. But this depends a lot on the situation/workflow, so I'm not sure there is even a general solution.

Accessibility of the file should not play a role, in my opinion. In any particular processing step, the current input file is guaranteed to be accessible; whether any earlier file is still accessible should not affect the meta data at the current step. Otherwise it's just confusing.

Where (by which tools) is the "spectra_data" information currently actually used?

**Comment by timosachsenberg** [comment]
e.g., PRIDE uses the information from the mzTab file to link spectra in the submitted mzML files to the PSMs.
I think you can't automatically get something you can visualize if there is just the information on the raw file but only the mzML files were submitted (e.g. full submission).
But I agree that probably having both types of information would make things easier (except for the cases where no raw files are annotated ;) ).
Maybe we should store both?

**Comment by hendrikweisser** [comment]
> e.g., PRIDE uses the information from the mzTab file to link spectra in the submitted mzML files to the PSMs.

That makes sense, but then you want to store the last mzML file there (the input to the step that produced the mzTab - or whichever file got converted to mzTab), not the first.

So would it make sense to store (at least) the "original" input file(s) - .raw if you have that information, otherwise the "earliest" you can get - and the most recent source (mzML) file?
(If you run an accurate mass search on a featureXML file, I think the "source" file that you want to reference in the ID results is that featureXML, not the mzML on which the feature detection was run.)

**Comment by timosachsenberg** [comment]
Storing only the last one doesn't work if the link to the original one is lost (e.g. if fractions are merged into a single mzML file).
The file in the primaryMSRun must match the one that is used in an experimental design file.
Otherwise, the mapping to samples or conditions also don't work.
Which brings us to the initial motivation to link data back to the sample/channel/condition/fraction whatever.

**Comment by hendrikweisser** [comment]
I get your point, but depending on how you merge, you may end up e.g. changing retention times (`FileMerger -rt_concat:...`), in which case only having the original files before the merge/change may also lead to discrepancies. If you do crazy stuff like spectral averaging, you may even create spectra that aren't present in your original files.

We could keep track of the whole analysis pipeline (see `DataProcessing`), but then we still have to pick a file from one step to write into an mzTab or whatever. I'm doubtful whether you can generally make this decision without user input.

**Comment by timosachsenberg** [comment]
Yes, I think this doesn't cover the crazy stuff. Julianus and me discussed this quite a bit and the way we do it seemed to be the only one that allows processing without user input while keeping track of the experimental design and associated files for all standard workflows.
Another problem is, that you can't take keep track of the whole analysis pipeline if you branch out.
But, yeah I agree that there are use cases where the raw file would also be interesting.

**Comment by hendrikweisser** [comment]
Okay, then let's keep it simple, in which case I agree with your previous suggestion of storing both the raw file (if available) and an mzML file. The raw file should anyway get passed through conversion steps in a `sourceFile` element in the mzML.

How do we make this change in `ProteinIdentification`? In the current design the mzML gets stored in a `spectra_data` meta value. Should we just add a `spectra_data_raw` element (or is that too Thermo-specific)?

By the way, my design in `IdentificationData` is that all ID processing steps get stored (in class `DataProcessingStep`), and each has a list of input files (specific to that step) and a list of primary files (that should get passed through).

**Comment by hendrikweisser** [comment]
The mzMLs are not stored in that case (except perhaps in the featureXMLs). That raw files that generated the mzMLs would be stored.

I don't know how experimental design is currently handled, but I don't understand why it should be tied to mzML. Isn't experimental design (in this context) about assigning samples and groups to files, which should be possible for any file type?

> TL;DR I think your ID data structure should explicitly provide a getPrimaryMSRun method as the other file formats do.

It does, but that would return the raw files instead of the mzMLs. I'm still not convinced of the necessity to store the mzMLs if they're only intermediate files (e.g. in the case of accurate mass search, there are no spectra to map IDs back to, so nothing to reference in the mzMLs).

**Comment by timosachsenberg** [comment]
You start a workflow with the spectra (mzML) and optionally some database file + the experimental design file.
So no featureXML or other files that you can reference in the experimental design file.

**Comment by hendrikweisser** [comment]
I see. So you basically use the mzML filename as the key to look up the experimental design annotation for a derived file. An alternative would be to store the annotations directly as meta data when you create files, but that's not how you implemented it now.
Another alternative would be to (optionally) formulate the experimental design in terms of raw files, which would make sense if your workflow includes the raw-to-mzML conversion. Maybe a good idea to accommodate experimentalists who work more with raw files directly?

Independent of that, you've convinced me that the easiest satisfactory solution is to add another field somewhere in IdentificationData. I'd prefer to keep it general and call it something like "experimental_design_reference", because in my view it's not really about mzML.

**Comment by timosachsenberg** [comment]
It basically serves two things:
- linking to the experimental design annotation (as discussed above)
- providing a sensible **default** for spectra visualization using quality control scripts/tools or visualization in TOPPView or other viewer

**Comment by poshul** [comment]
@hendrikweisser are you okay with closing this now?

**Comment by hendrikweisser** [comment]
Thanks for the reminder, Sam.
